omroneipApp/src/drvOmroneip.cpp

```diff
@@ -0,0 +1,1882 @@
+{
+  std::vector<std::string> row = rawMap.at(structName);
+  std::vector<std::string> expandedRow;
+  for (std::string dtype : row)
+  {
+    if (dtype =="UINT" || dtype =="INT" || dtype =="WORD" || dtype =="BOOL" || dtype =="UDINT" || dtype =="DINT" || dtype =="REAL" || dtype =="DWORD" || dtype =="ULINT" || dtype =="LINT" || dtype == "LREAL" || dtype =="LWORD" || dtype.substr(0,6) == "STRING")
+    {
+      expandedRow.push_back(dtype);
+    }
+    else
+    {
+      // If we get here, then we assume that we have a structure, but it could be a typo
+      try
+      {
+        //recursively call this function to get all of the embedded datatypes
+        std::vector<std::string> embeddedDtypes = expandStructs(rawMap, dtype);
+        expandedRow.insert(std::end(expandedRow), std::begin(embeddedDtypes), std::end(embeddedDtypes));
+      }
+      catch (...)
+      {
+        std::cout<<"Failed to find the standard datatype: " << dtype << ". Definition for " << structName << " and its dependents failed" << std::endl;
+      }
+    }
+  }
+  return expandedRow;
+}
+
+size_t drvOmronEIP::findOffsetsRecursive(std::unordered_map<std::string, std::vector<std::string>> const& rawMap, std::unordered_map<std::string, std::vector<std::string>> const& expandedMap, std::string structName, std::unordered_map<std::string, std::vector<int>>& structMap, std::string parentNextDtype)
+{
+  std::vector<int> newRow = {0};
+  size_t offSetsCounted = 0;
+  std::vector<std::string> row = rawMap.at(structName);
+  std::vector<std::string> expandedRow = expandedMap.at(structName);
+  size_t lastOffset = 0;
+  bool structInvalid = false;
+  size_t thisOffset = 0;
+  size_t dtypeSize = 0;
+  std::string nextDtype = "none";
+  size_t consequitiveDtypes = 1;
+  size_t padSize = 0;
+  size_t index = 0;
+  size_t nextSize =0;
+  size_t consequitiveBools = 0;
+  bool found = false;
+  // loop through each datatype and find its byte size
+  for (std::string dtype : row)
+  {
+    index++;
+    {
+      // Here we are searching for the next standard dtype so we can calculate if there will be padding
+      if (index < row.size()){
+        nextDtype = row[index];
+        nextDtype = findNextDtype(rawMap, nextDtype);
+        if (nextDtype == "no dtypes left") {
+          // Case where the next dtype is a custom struct but it contains no standard dtypes
+          continue;
+        }
+        else if ( nextDtype == "structName is neither a structure or a standard dtype") {
+          std::cout<<"Invalid dtype/struct name: " << structName <<std::endl;
+          return -1;
+        }
+      }
+      else {
+        // Padding may be required at the end of the structure
+        nextDtype=parentNextDtype;
+      }
+    }
+
+    if (dtype =="UINT" || dtype =="INT" || dtype =="WORD" || dtype =="BOOL")
+      dtypeSize=2;
+    else if (dtype =="UDINT" || dtype =="DINT" || dtype =="REAL" || dtype =="DWORD")
```

Comment:
are `UINT` and `INT` really only 16bits? Presumably that's the PLCs datatypes...?

Reply:
Yes they are